### PR TITLE
[RM-1002] Add a new list

### PR DESCRIPTION
**Pull request by vxduc9999**
When click add list button then create new list to list data
and set idList in global editting with id new list

I modify the personal list file because on clicking add list
then show new list in it and auto focus to enter the name
To save with 2 ways:
 + Enter key
 + Click anywhere in the web page using blur focus technique

I use icon from google and font format for all text,
I push data at web start and add global editing
to make it easy to edit names. Modify css for better sidebar

References: RM-1002

### CODE REVIEW

File: public/index.html [hunk 1]
```diff
@@ -5,6 +5,10 @@
     <meta http-equiv="X-UA-Compatible" content="IE=edge">
     <meta name="viewport" content="width=device-width,initial-scale=1.0">
     <link rel="icon" href="<%= BASE_URL %>favicon.ico">
+    <link href="https://fonts.googleapis.com/css2?family=Material+Icons" rel="stylesheet">
+    <link rel="preconnect" href="https://fonts.googleapis.com">
+    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
```
Comment (phuphamphong97) [inline public/index.html]: Why we need these links?

File: src/components/SideBar/AddListButton.vue [hunk 2]
```diff
@@ -1,3 +1,42 @@
 <template>
-  <button>AddListButton</button>
+  <div class="add-list-button" @click="AddList">
+    <i class="material-icons" id="icon">
+      add_circle_outline
+    </i>
+    <button class="button">Add List</button>
+  </div>
```
Comment (phuphamphong97) [inline src/components/SideBar/AddListButton.vue]: Why don't we put the icon inside the button? Remember to think of semantics in HTML: a button should be a _real_ `<button>`, instead of a _fake_ one using a `<div>` with its `@click` event.

File: src/components/SideBar/AddListButton.vue [hunk 3]
```diff
@@ -1,3 +1,42 @@
...
 </template>
+
+<script>
+export default {
+  name: "AddListButton",
+  methods: {
+    AddList() {
```
Comment (phuphamphong97) [inline src/components/SideBar/AddListButton.vue]: Function name should be in `camelCase`.

File: src/components/SideBar/AddListButton.vue [hunk 4]
```diff
@@ -1,3 +1,42 @@
...
+      const endId = this.$store.listsData[this.$store.listsData.length - 1].id;
```
Comment (phuphamphong97) [inline src/components/SideBar/AddListButton.vue]: If `this.$store.listsData` is used multiple times, we can set it to a computed property to avoid code duplication.
E.g:
```
computed: {
    listsData() {
        return this.$store.listsData;
    }
}
```

Also, this line can cause an error when `listData` is empty: `id of undefined` (or something like that).

File: src/components/SideBar/AddListButton.vue [hunk 5]
```diff
@@ -1,3 +1,42 @@
...
+      this.$store.listsData.push({
+        id: endId + 1,
+        name: "New a list" + " (" + String(endId + 1) + ")",
```
Comment (phuphamphong97) [inline src/components/SideBar/AddListButton.vue]: - "New list", not "New a list"
- Should use template string in ES6: 
```
`New list (${endId + 1})`
```

File: src/components/SideBar/AddListButton.vue [hunk 6]
```diff
@@ -1,3 +1,42 @@
...
+      });
+      this.$store.editting.idList = endId + 1;
```
Comment (phuphamphong97) [inline src/components/SideBar/AddListButton.vue]: - Naming suggestion: `lastListId` instead of `endId`
- Should create a separate const for the new list id right after getting the last list id, so that the context below is clearer (compared to using `endId + 1`)

File: src/components/SideBar/AddListButton.vue [hunk 7]
```diff
@@ -1,3 +1,42 @@
...
+};
+</script>
+
+<style scoped>
+.add-list-button {
+  display: flex;
+  padding-left: 5px;
+}
+
+#icon {
```
Comment (phuphamphong97) [inline src/components/SideBar/AddListButton.vue]: Avoid styling an element based on its id, use class selectors instead.

File: src/components/SideBar/SeachTask.vue [hunk 8]
```diff
@@ -6,6 +6,6 @@
 
 <style scoped>
 .search {
-  height: 10%;
+  height: 5%;
```
Comment (phuphamphong97) [inline src/components/SideBar/SeachTask.vue:9]: This ticket is not related to the search feature, so this change should be reverted.

File: src/components/SideBar/SideBar.vue [hunk 9]
```diff
@@ -25,8 +25,9 @@ export default {
 
 <style scoped>
 .side-bar {
-  background-color: rgb(192, 187, 187);
+  background-color: rgba(230, 245, 247, 0.5);
   width: 30%;
-  padding: 10px 0 0 10px;
+  min-width: 10%;
+  padding: 20px 10px 0 10px;
```
Comment (phuphamphong97) [inline src/components/SideBar/SideBar.vue:31]: All formatting changes should be in a separate commit since they're not related to adding a new list.

File: src/store.js [hunk 10]
```diff
@@ -11,4 +14,12 @@ export default {
     this.tasksData = [...tasks];
     return this.tasksData;
   },
+  countAllTasksByIdList(idList) {
```
Comment (phuphamphong97) [inline src/store.js]:
```suggestion
  countAllTasksByListId(listId) {
```

File: src/store.js [hunk 11]
```diff
@@ -3,6 +3,9 @@ import { lists, tasks } from "@/data";
 export default {
   listsData: [],
   tasksData: [],
+  editting: {
+    idList: -1,
```
Comment (phuphamphong97) [inline src/store.js]: - Naming suggestion: `listId` instead of `idList`
- Initial value should be `null`, so when doing a comparison, just write
`if (!listId)` or `if (listId)`
instead of
`if (listId === -1)` or `if (listId !== -1)`

File: src/store.js [hunk 12]
```diff
@@ -11,4 +14,12 @@ export default {
     this.tasksData = [...tasks];
     return this.tasksData;
   },
+  countAllTasksByIdList(idList) {
+    return this.tasksData.filter((task) => task.id_list === idList).length;
+  },
+  editListName(idList, listName) {
+    const index = this.listsData.findIndex((list) => list.id === idList);
+    this.listsData[index].name = listName;
```
Comment (phuphamphong97) [inline src/store.js:26]: Should check for the case ` index = -1` (no item found with the given id).

File: src/components/SideBar/PersonalLists.vue [hunk 13]
```diff
@@ -1,46 +1,175 @@
...
-      </div>
-      <div class="personal-lists-detail-right">
-        <p>0</p>
-        <p>kebab_menu</p>
+      <div
+        class="personal-lists-detail-element"
+        v-for="list in $lists"
+        :key="list.id"
+        :class="{
+          'list-selected': listSelected == list.id,
```
Comment (phuphamphong97) [inline src/components/SideBar/PersonalLists.vue]: Always remember to use `===` in comparisons.

File: src/components/SideBar/PersonalLists.vue [hunk 14]
```diff
@@ -1,46 +1,175 @@
...
+        }"
+        @click="listSelected = list.id"
+      >
+        <div class="personal-lists-detail-left">
+          <span class="material-icons" id="icon">
```
Comment (phuphamphong97) [inline src/components/SideBar/PersonalLists.vue]: Consider to use only one tag for icons: `<span>` or ` <i>`.

File: src/components/SideBar/PersonalLists.vue [hunk 15]
```diff
@@ -1,46 +1,175 @@
...
+  directives: {
+    focus: {
+      inserted(el) {
+        Vue.nextTick(() => el.focus());
+      },
+    },
+  },
+  data() {
+    return {
+      listSelected: -1,
```
Comment (phuphamphong97) [inline src/components/SideBar/PersonalLists.vue]: - Naming suggestion: `selectedListId`, not `listSelected`
- It should be moved to the global store as we might need it later for routing or getting its tasks from the right panel.
- The initial value should be `null`, as I explained in the previous comment.

File: src/components/SideBar/PersonalLists.vue [hunk 16]
```diff
@@ -1,46 +1,175 @@
...
+    };
+  },
+  computed: {
+    $lists: {
+      get() {
+        return this.$store.listsData;
+      },
+    },
```
Comment (phuphamphong97) [inline src/components/SideBar/PersonalLists.vue]: If a computed property doesn't have the setter function, it can be shortened as below:
```
$lists() {
     return this.$store.listsData;
}

File: src/components/SideBar/PersonalLists.vue [hunk 17]
```diff
@@ -1,46 +1,175 @@
...
+    $edittingList: {
+      get() {
+        return this.$store.editting.idList;
+      },
+    },
+  },
+  methods: {
+    countAllTasksByIdList(idList) {
+      return this.$store.countAllTasksByIdList(idList);
+    },
```
Comment (phuphamphong97) [inline src/components/SideBar/PersonalLists.vue]: Why don't we just call `$store.countAllTasksByIdList(idList)` directly on the template, rather than declaring an additional method just to call another one from the store?

File: src/components/SideBar/PersonalLists.vue [hunk 18]
```diff
@@ -1,46 +1,175 @@
...
+          (list) => list.name === listName
+        );
+        if (
+          listExist === undefined ||
+          (listExist.name === listName && listExist.id === idList)
+        ) {
+          this.$store.editListName(idList, listName);
+        }
+      }
+    },
```
Comment (phuphamphong97) [inline src/components/SideBar/PersonalLists.vue]: - It would be better if we show warnings when the user leaves the input blank or the new list name is duplicated with the existing one.
- Don't mutate the function input (`listName`).
- Irrelevant conditions should be kept separate.

E.g:

```
editListName(listId, $event) {
  const listName = $event.target.value.toString().trim();
  if (!listName) {
    return console.log("List name required!");
  }

  const duplicated = !!this.$lists.find(
    (list) => list.id !== listId && list.name === listName
  );
  if (duplicated) {
    return console.log("Duplicated name!");
  }

  this.$store.editListName(listId, listName);
}
```

**Review by phuphamphong97 — CHANGES_REQUESTED** [review]
Some hopefully helpful comments on your changes.

File: public/index.html (hunk 1 above)
Comment (vxduc9999) [reply public/index.html]: I copied it in the docs page, I try to ignore it and it still works. I've updated

File: src/components/SideBar/AddListButton.vue (hunk 2 above)
Comment (vxduc9999) [reply src/components/SideBar/AddListButton.vue]: I updated, please check again

File: src/components/SideBar/AddListButton.vue (hunk 3 above)
Comment (vxduc9999) [reply src/components/SideBar/AddListButton.vue]: Updated

File: src/components/SideBar/AddListButton.vue (hunk 4 above)
Comment (vxduc9999) [reply src/components/SideBar/AddListButton.vue]: Updated!

File: src/components/SideBar/AddListButton.vue (hunk 5 above)
Comment (vxduc9999) [reply src/components/SideBar/AddListButton.vue]: Updated!

File: src/components/SideBar/AddListButton.vue (hunk 6 above)
Comment (vxduc9999) [reply src/components/SideBar/AddListButton.vue]: Updated

File: src/components/SideBar/AddListButton.vue (hunk 7 above)
Comment (vxduc9999) [reply src/components/SideBar/AddListButton.vue]: Updated

File: src/components/SideBar/PersonalLists.vue (hunk 13 above)
Comment (vxduc9999) [reply src/components/SideBar/PersonalLists.vue]: Updated

File: src/components/SideBar/PersonalLists.vue (hunk 14 above)
Comment (vxduc9999) [reply src/components/SideBar/PersonalLists.vue]: Updated!

File: src/components/SideBar/PersonalLists.vue (hunk 15 above)
Comment (vxduc9999) [reply src/components/SideBar/PersonalLists.vue]: Updated!

File: src/components/SideBar/PersonalLists.vue (hunk 16 above)
Comment (vxduc9999) [reply src/components/SideBar/PersonalLists.vue]: Updated!

File: src/components/SideBar/PersonalLists.vue (hunk 17 above)
Comment (vxduc9999) [reply src/components/SideBar/PersonalLists.vue]: Updated!

File: src/components/SideBar/PersonalLists.vue (hunk 18 above)
Comment (vxduc9999) [reply src/components/SideBar/PersonalLists.vue]: Updated!

File: src/store.js (hunk 11 above)
Comment (vxduc9999) [reply src/store.js]: Updated!

File: src/store.js (hunk 10 above)
Comment (vxduc9999) [reply src/store.js]: Updated!

File: src/store.js (hunk 12 above)
Comment (vxduc9999) [reply src/store.js:26]: Updated!

File: src/components/SideBar/SeachTask.vue (hunk 8 above)
Comment (vxduc9999) [reply src/components/SideBar/SeachTask.vue:9]: Updated!

File: src/components/SideBar/SideBar.vue (hunk 9 above)
Comment (vxduc9999) [reply src/components/SideBar/SideBar.vue:31]: I move it to commit Update css. Please check!

File: src/store.js [hunk 19]
```diff
@@ -11,4 +15,13 @@ export default {
     this.tasksData = [...tasks];
     return this.tasksData;
   },
+  countAllTasksBylistId(listId) {
```
Comment (phuphamphong97) [inline src/store.js]:
```suggestion
  countAllTasksByListId(listId) {
```

File: package.json [hunk 20]
```diff
@@ -9,9 +9,10 @@
   },
   "dependencies": {
     "core-js": "^3.6.5",
+    "nanoid": "^3.1.23",
     "vue": "^2.6.11",
-    "vue-stash": "^2.0.1-beta",
-    "vue-router": "^3.2.0"
+    "vue-router": "^3.2.0",
+    "vue-stash": "^2.0.1-beta"
```
Comment (phuphamphong97) [inline package.json]: Should revert this change since it's not related to this ticket.

File: src/components/SideBar/AddListButton.vue [hunk 21]
```diff
@@ -1,3 +1,45 @@
 <template>
-  <button>AddListButton</button>
+  <button class="button" @click="addList">
+    <i class="add-list-icon material-icons"> add_circle_outline </i
+    ><span>Add List</span>
```
Comment (phuphamphong97) [inline src/components/SideBar/AddListButton.vue]:
```suggestion
    <i class="add-list-icon material-icons"> add_circle_outline </i>
    <span>Add List</span>
```

File: src/components/SideBar/PersonalList.vue [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+      const duplicated = !!this.lists.find(
+        (list) => list.id !== listId && list.name === listName
+      );
+
+      if (duplicated) {
+        return alert("Duplicated name!");
+      }
+
+      const isSuccess = this.$store.editListName(listId, listName);
+      if (isSuccess) return alert("There is something wrong!");
```
Comment (phuphamphong97) [inline src/components/SideBar/PersonalList.vue]: `editListName()` doesn't return boolean in all possible cases, so this variable type should be checked.
And, such a magical logic when showing an error message after a successful action!

File: package.json (hunk 20 above)
Comment (vxduc9999) [reply package.json]: Updated!

File: src/components/SideBar/AddListButton.vue (hunk 21 above)
Comment (vxduc9999) [reply src/components/SideBar/AddListButton.vue]: Updated!

File: src/store.js (hunk 19 above)
Comment (vxduc9999) [reply src/store.js]: Updated!

File: src/components/SideBar/PersonalList.vue (hunk 22 above)
Comment (vxduc9999) [reply src/components/SideBar/PersonalList.vue]: Updated!

File: src/store.js [hunk 23]
```diff
@@ -11,4 +15,14 @@ export default {
     this.tasksData = [...tasks];
     return this.tasksData;
   },
+  countAllTasksByListId(listId) {
+    return this.tasksData.filter((task) => task.id_list === listId).length;
+  },
+  editListName(listId, listName) {
+    const index = this.listsData.findIndex((list) => list.id === listId);
+    if (index === -1) return false;
```
Comment (phuphamphong97) [inline src/store.js]:
```suggestion
    if (index === -1) {
        return false;
    }
```

**Review by phuphamphong97 — APPROVED** [review]
Just one more comment to resolve then it looks good to me!
You need to practice more to improve the commit messages.

Please squash your fixup commits before merging.

File: src/store.js (hunk 23 above)
Comment (vxduc9999) [reply src/store.js]: Updated!
